### PR TITLE
Add Feature: create initial snapshot of VM after clean-up

### LINKED ISSUE #20: Create initial snapshot after cleanup
#### Description

Create an initial snapshot of the newly created VM after the clean-up step.

#### Use Case(s)

- if you need to return frequently to a clean state of the VM (right after the initial setup) e.g. to clear files / malware / misconfigurations, the initial snapshot to return to can be shipped right with the VM so the user doesn't have to create it on its own and potentially forget it.

#### Potential configuration

1. add a create snapshot step after the `StepCleanVMX` for the respective builder
```
&vmwcommon.StepCreateSnapshot{
	SnapshotName: &b.config.SnapshotName,
},
```
2. add `SnapshotName string `mapstructure:"snapshot_name" required:"false"`` to the builders config to be able to specify the snapshots name in the config
3. add a `builder/vmware/common/step_create_snapshot.go` which creates a snapshot if the name is provided in the config
```
type StepCreateSnapshot struct {
	SnapshotName *string
}
```
4. add a `CreateSnapshot` method inside the drivers which calls `vmrun.exe` or equivalent tool on linux / macos to create a snapshot
5. add Tests

### PR DESCRIPTION
**Pull request by 0xFaul**
* extended drivers by `CreateSnapshot` method to create snapshots
* added additional step to builders to create initial snapshot of the finished VM (after clean-up)

simply set the `"snapshot_name"` field inside the builder's config to create an initial snapshot (if not set, snapshot creation will be skipped)

Closes #20

### CODE REVIEW

File: builder/vmware/common/step_create_snapshot.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+		if err := driver.CreateSnapshot(vmFullPath, *s.SnapshotName); err != nil {
+			err := fmt.Errorf("Error creating snapshot: %s", err)
+			state.Put("error", err)
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+		state.Put("snapshot_created", true)
+	} else {
+		state.Put("snapshot_skipped", true)
+	}
```
[the remark this answers is not in the document]
Comment (0xFaul) [reply builder/vmware/common/step_create_snapshot.go:40]: those are at the moment mainly for testing and to have the state as an internal checklist, but are not necessarily needed.

[the remark this answers is not in the document]
Comment (sylviamoss) [reply builder/vmware/common/step_create_snapshot.go:40]: Oh sorry, I didn't notice it was being used for testing. Sounds good!

File: builder/vmware/iso/config.go [hunk 2]
```diff
@@ -75,6 +75,9 @@ type Config struct {
 	// non-functional. See below for more information. For basic VMX
 	// modifications, try `vmx_data` first.
 	VMXTemplatePath string `mapstructure:"vmx_template_path" required:"false"`
+	// This is the name of the initial snapshot created after provisioning and cleanup.
```
Comment (sylviamoss) [inline builder/vmware/iso/config.go:78]: To update the generated docs from this you should use `go generate ./...`

File: main.go [hunk 3]
```diff
@@ -12,7 +12,7 @@ import (
 
 var (
 	// Version is the main version number that is being run at the moment.
-	Version = "0.0.2"
+	Version = "0.0.3"
```
Comment (sylviamoss) [inline main.go:15]: This one is still two, we update once we release the 0.0.2
```suggestion
	Version = "0.0.2"
```

**Review by sylviamoss — APPROVED** [review]
This is looking good 👍🏼 
It is missing only the generated docs which you get by running `go generate ./...`

File: builder/vmware/iso/config.go (hunk 2 above)
Comment (0xFaul) [reply builder/vmware/iso/config.go:78]: done :)

**Comment by sylviamoss** [comment]
Thank you!

**Pull request merged**